File: tso_kernel/operators.py
"""
Geometric operators for TSO.

Double Mapping, topological expansion, and the Inverse Motor projection.
All operations are local, purely geometric — no global gradient.
"""
import numpy as np


class TopographicOperator:
    """
    Static methods for the core TSO geometric transforms.
    """

    @staticmethod
    def double_mapping(vectors, concepts, context, d):
        """
        Project conflicting concepts into disjoint subspaces.

        Each concept gets its own d-dimensional block within a k*d space.
        The context is replicated across all blocks. This guarantees
        that the dot product between exclusive concepts is zero
        while context similarities are preserved.

        Args:
            vectors: dict mapping label -> 1D array (length d)
            concepts: list of concept labels to orthogonalize
            context: label of the context element
            d: intrinsic dimension

        Returns:
            new_vectors: dict mapping label -> expanded array (length k*d)
            nd: new dimension
        """
        k = len(concepts)
        nd = k * d
        new_vecs = {}

        for c in concepts:
            zp = np.zeros(nd, dtype=np.float32)
            i = concepts.index(c)
            zp[i*d:(i+1)*d] = vectors[c]
            new_vecs[c] = zp

        ctx_zp = np.zeros(nd, dtype=np.float32)
        for i in range(k):
            ctx_zp[i*d:(i+1)*d] = vectors[context]
        new_vecs[context] = ctx_zp

        return new_vecs, nd

    @staticmethod
    def inverse_motor(state, W_proj, embeddings, cluster_mask):
        """
        Project SNN state to embedding space and select best word
        within the target cluster.

        Args:
            state: SNN state vector
            W_proj: projection matrix (state_dim x embed_dim)
            embeddings: word embedding matrix (vocab_size x embed_dim)
            cluster_mask: boolean array (vocab_size,) for the target cluster

        Returns:
            best_idx: index of best matching word
            best_cos: cosine similarity score
        """
        proj = state @ W_proj
        norms = np.linalg.norm(proj) * np.linalg.norm(embeddings, axis=1) + 1e-8
        cosines = (embeddings @ proj) / norms
        cosines[~cluster_mask] = -1.0
        best_idx = int(np.argmax(cosines))
        return best_idx, float(cosines[best_idx])

    @staticmethod
    def expand_topology(W_exc, W_inh, dormant_threshold=0.01):
        """
        Phase 2: detect edge saturation and flag dormant neurons
        for recruitment. Returns indices of saturated edges.
        """
        n = W_exc.shape[0]
        saturated = []
        for i in range(n):
            for j in range(n):
                if abs(W_exc[i, j]) > 1.0 - dormant_threshold:
                    saturated.append((i, j))
        return saturated

File: tso-kernel/src/operators.rs
use ndarray::Array1;
use std::collections::HashMap;

pub struct TopographicOperator;

impl TopographicOperator {
    /// Strict Double Mapping (Lemma 1).
    /// Projects conflicting concepts into fully disjoint subspaces.
    /// Returns (new_vectors, new_dimension).
    pub fn double_mapping(
        vectors: &HashMap<String, Array1<f64>>,
        concepts: &[String],
        context: &str,
        d: usize,
    ) -> (HashMap<String, Array1<f64>>, usize) {
        let k = concepts.len();
        let nd = k * d;
        let mut new_vecs = HashMap::new();

        for (i, c) in concepts.iter().enumerate() {
            let mut zp = Array1::zeros(nd);
            if let Some(v) = vectors.get(c) {
                for j in 0..d {
                    zp[i * d + j] = v[j];
                }
            }
            new_vecs.insert(c.clone(), zp);
        }

        let mut ctx_zp = Array1::zeros(nd);
        if let Some(ctx_v) = vectors.get(context) {
            for i in 0..k {
                for j in 0..d {
                    ctx_zp[i * d + j] = ctx_v[j];
                }
            }
        }
        new_vecs.insert(context.to_string(), ctx_zp);

        (new_vecs, nd)
    }

    /// Soft Double Mapping with residual shared space (Lemma 1 v2).
    /// Instead of strict orthogonality, preserves shared features
    /// between exclusive concepts via an alpha coefficient.
    pub fn soft_double_mapping(
        vectors: &HashMap<String, Array1<f64>>,
        concepts: &[String],
        context: &str,
        d: usize,
        alpha: f64,
    ) -> (HashMap<String, Array1<f64>>, usize) {
        let k = concepts.len();
        let nd = k * d;
        let mut new_vecs = HashMap::new();

        for (i, c) in concepts.iter().enumerate() {
            let mut zp = Array1::zeros(nd);
            if let Some(v) = vectors.get(c) {
                for j in 0..d {
                    zp[i * d + j] = v[j];
                }
                for other_idx in 0..k {
                    if other_idx != i {
                        for j in 0..d {
                            zp[other_idx * d + j] = v[j] * alpha;
                        }
                    }
                }
            }
            new_vecs.insert(c.clone(), zp);
        }

        let mut ctx_zp = Array1::zeros(nd);
        if let Some(ctx_v) = vectors.get(context) {
            for i in 0..k {
                for j in 0..d {
                    ctx_zp[i * d + j] = ctx_v[j];
                }
            }
        }
        new_vecs.insert(context.to_string(), ctx_zp);

        (new_vecs, nd)
    }

    /// Inverse Motor: project SNN state to embedding space and select best word.
    ///
    /// Returns (best_index, cosine_similarity).
    pub fn inverse_motor(
        state: &Array1<f64>,
        w_proj: &Array1<f64>,
        embeddings: &[Array1<f64>],
        cluster_mask: &[bool],
    ) -> (usize, f64) {
        let state_dim = state.len();
        let embed_dim = w_proj.len() / state_dim;
        let proj_vec = apply_projection(state, w_proj, state_dim, embed_dim);

        let mut best_idx = 0;
        let mut best_cos = -2.0;

        for (i, emb) in embeddings.iter().enumerate() {
            if !cluster_mask[i] {
                continue;
            }
            let dot = emb.dot(&proj_vec);
            let na = emb.dot(emb).sqrt();
            let nb = proj_vec.dot(&proj_vec).sqrt();
            let cos = dot / (na * nb + 1e-8);
            if cos > best_cos {
                best_cos = cos;
                best_idx = i;
            }
        }

        (best_idx, best_cos)
    }
}

fn apply_projection(state: &Array1<f64>, w_proj: &Array1<f64>, state_dim: usize, embed_dim: usize) -> Array1<f64> {
    let mut result = Array1::zeros(embed_dim);
    for j in 0..embed_dim {
        let mut dot = 0.0;
        for i in 0..state_dim {
            dot += state[i] * w_proj[i * embed_dim + j];
        }
        result[j] = dot;
    }
    result
}
